File: TABLE/table.c
/**
 * Author: 		    Jeroen Overschie
 * S-number:		s2995697
 * Date:			24-02-2018
 * Challenge:	    TABLE
 * 
 * The input for this program are three numbers n, m and k. The program
 * then computes the k-th largest number in the multiplication table of
 * up to n * m.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define __FILENAME__ \
    (strrchr(__FILE__, '/') ? strrchr(__FILE__, '/') + 1 : __FILE__)
#ifdef TEST
#define DEBUG_PRINT(fmt, ...)                                                   \
    fprintf(stderr, "%s %s %s:%d:%s(): " fmt, __DATE__, __TIME__, __FILENAME__, \
        __LINE__, __func__, ##__VA_ARGS__)
#define TRACE_PRINT(fmt, ...) fprintf(stderr, fmt, ##__VA_ARGS__)
#else
#define DEBUG_PRINT(fmt, ...)
#define TRACE_PRINT(fmt, ...)
#endif

#define MIN(a, b) (((a) < (b)) ? (a) : (b))
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

typedef unsigned long int lint; // shorthand

lint getKthLargestNumber(lint n, lint m, lint k)
{
    lint smallest = 1, largest = m * n;

    while (smallest < largest) {
        lint newMin = smallest + (largest - smallest) / 2;

        if (!isSufficient(n, m, k, newMin)) {
            smallest = newMin + 1;
        } else {
            largest = newMin;
        }
    }

    return smallest;
}

int isSufficient(lint n, lint m, lint k, lint x)
{
    lint count = 0;
    for (lint i = 1; i <= m; i++) {
        count += MIN(x / i, n);
    }
    return count >= k;
}

int main(int argc, char* argv[])
{
// Redirect stdin to a test file in case of TEST environment
#ifdef TEST
    freopen("./TABLE/2.in", "r", stdin);
#endif
    DEBUG_PRINT("TABLE\n");

    lint n, m, k;
    scanf("%lu %lu %lu\n", &n, &m, &k);
    DEBUG_PRINT("Scanned numbers:\n");
    TRACE_PRINT("%d\n", n);
    TRACE_PRINT("%d\n", m);
    TRACE_PRINT("%d\n", k);

    lint l = getKthLargestNumber(n, m, k);
    printf("%lu\n", l);

    return 0;
}

File: TABLE/table.py
# Author: 		Jeroen Overschie
# S-number:		s2995697
# Date:			25-02-2018
# Challenge:	TABLE
# 
# The input for this program are three numbers n, m and k. The program
# then computes the k-th largest number in the multiplication table of
# up to n * m.

import logging
import os
import sys
import math

# LOGGING
DEBUG_LEVEL = logging.DEBUG if "TEST" in os.environ else logging.CRITICAL
logging.basicConfig(stream=sys.stderr, level=DEBUG_LEVEL)
# REDIRECT STDIN
if "TEST" in os.environ:
    old_stdin = sys.stdin  # save it, in case we need to restore it
    sys.stdin = open('./TABLE/2.in')

logging.info("TABLE")

def getKthLargestNumber(n, m, k):
    mplus = m + 1
    def isSufficient(x):
        count = 0
        for i in range(1, mplus):  # would be xrange in Python 2.x
            # use Python 3.x's `//` operator to obtain int
            count += min(x // i, n)
        return count >= k

    smallest, largest = 1, m * n
    while smallest < largest:
        # use Python 3.x's `//` operator to obtain int
        mi = (smallest + largest) // 2
        if not isSufficient(mi):
            smallest = mi + 1
        else:
            largest = mi
    return smallest

n, m, k = [int(x) for x in input().split()]

logging.debug('n = %d', n)
logging.debug('m = %d', m)
logging.debug('k = %d', k)

# The answer is printed normally
print(math.floor(getKthLargestNumber(n, m, k)))
